dev/Kozani/KozaniRemoteManager/KozaniDvcServer.cpp

```diff
@@ -0,0 +1,734 @@
+        HRESULT hr = LOG_IF_FAILED(ProcessActivateAppRequest(pdu, appActivatedAndTracked, errorMessage));
+
+        // If ProcessActivateAppRequest failed but appActivatedAndTracked is true, it means there was failure after the app was activated and 
+        // its process lifetime tracked. Such failure is benign and we should not send failure ActivateAppResult to client.
+        if (FAILED(hr) && !appActivatedAndTracked)
+        {
+            SendActivateAppResult(pdu.activity_id(), hr, 0, false, errorMessage);
+
+            // ERROR_ALREADY_EXISTS is due to the activity Id being the same as an existing activity already tracked in the m_activityMap. 
+            // Do not remove it from the map. 
+            if (hr != HRESULT_FROM_WIN32(ERROR_ALREADY_EXISTS))
+            {
+                auto lock{ g_connectionManager.m_activityMapLock.lock_exclusive() };
+                if (g_connectionManager.m_closing)
+                {
+                    return;
+                }
+
+                g_connectionManager.m_activityMap.erase(pdu.activity_id());
+            }
+        }
+    }
+    CATCH_LOG_RETURN()
+
+    HRESULT KozaniDvcServer::ProcessActivateAppRequest(
+        _In_ Dvc::ProtocolDataUnit& pdu, 
+        _Out_ bool& appActivatedAndTracked, 
+        _Out_ std::string& errorMessage) noexcept try
+    {
+        appActivatedAndTracked = false;
+        errorMessage.clear();
+
+        Dvc::ActivateAppRequest activateAppRequest;
+        if (!activateAppRequest.ParseFromString(pdu.data()))
+        {
+            errorMessage = "Failed to parse ActivateAppRequest message";
+            RETURN_WIN32(ERROR_INVALID_DATA);
+        }
+
+        if (!IsActivationKindSupported(activateAppRequest.activation_kind()))
+        {
+            errorMessage = "Unsupported activation kind: " + std::to_string(activateAppRequest.activation_kind());
+            RETURN_WIN32(ERROR_NOT_SUPPORTED);
+        }
+
+        {
+            // Acquire lock before adding the activity Id to the m_activityMap but do not hold the lock while activating the app as 
+            // that can take some time and we do not want one app activation to block others.
+            auto lock{ g_connectionManager.m_activityMapLock.lock_exclusive() };
+            RETURN_HR_IF(HRESULT_FROM_WIN32(ERROR_INVALID_STATE), g_connectionManager.m_closing);
+
+            auto inserted{ g_connectionManager.m_activityMap.emplace(pdu.activity_id(), AppActivationInfo(pdu.activity_id())) };
+            if (!inserted.second)
+            {
+                errorMessage = "Activity ID of the new activate app request is already used for a previous request. Each new request must use a new activity ID.";
+                RETURN_WIN32(ERROR_ALREADY_EXISTS);
+            }
+        }
+
+        wil::com_ptr<IApplicationActivationManager> aam;
+        HRESULT hrCoCreateInstance{ CoCreateInstance(CLSID_ApplicationActivationManager, nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&aam)) };
```

Comment:
It will NOT return an HRESULT:

/** constructs a COM object using an CLSID on a specific interface or IUnknown.
    Note, failures are reported as a null result, the HRESULT is lost. */
    template<typename Interface = IUnknown>
    wil::com_ptr_nothrow<Interface> CoCreateInstanceNoThrow(REFCLSID rclsid, DWORD dwClsContext = CLSCTX_INPROC_SERVER) WI_NOEXCEPT
    {
        return CoCreateInstance<Interface, err_returncode_policy>(rclsid, dwClsContext);
    }

Reply:
Ack! That's...just sad. That HRESULT routinely has at least 2 significantly useful values for radically different issues. Grrrrrr.

Yeah. Don't use that :P I'll ask WIL folks if there's something better we're overlooking